### ISSUE TITLE
imfile stopped working after logrotate again

### ISSUE THREAD

**Issue by waster** [rsyslog/rsyslog#3618]
We use rsyslog to send logs via RELP for icecast logs with imfile module.

After some time (in 1-2 weeks) it can silently stop to monitor some files and send most of the RELP messages and recover after next log rotate. No any warnings or errors are stored in rsyslog log file. 

We observe such persistent problem last few releases after 8.31.0.
Log rotation is done internally via Iceast with "<logsize></logsize>" setting in configuration file.

### Environment
- rsyslog version: rsyslogd  8.1903.0
- platform: Ubuntu 16.04
- rsyslog.conf:

```
module(load="imfile" mode="polling")
module(load="omrelp")

input(type="imfile"
    File="/home/icecast/*/log/access.log"
    addMetadata="on"
    Tag="remoteIcecast"
)

input(type="imfile"
    File="/home/icecast/*/log/playlist.log"
    addMetadata="on"
    Tag="remoteIcecast"
)
...
ruleset(name="sendToStatServer") {
    action(type="omrelp"
        name="sendToStatServer"
        Target="xxx.xxx.xxx"
        Port="10514"
        Template="StatFormat"
        queue.type="fixedArray"
        queue.size="10000"
        queue.filename="q_sendToStatserver"
        queue.highwatermark="9000"
        queue.lowwatermark="100"
        queue.maxdiskspace="500m"
        queue.saveonshutdown="on"
        queue.dequeueBatchSize="100"
        action.resumeRetryCount="-1"
        action.reportSuspension="on"
        action.reportSuspensionContinuation="on"
        action.resumeInterval="10"
    )
}
```
Will that configuration work stable with wildcards and polling mode? What should we use to have production stable monitoring of files, _inotify + statefilealwaysremove_?

**Comment by rgerhards** [comment]
do you use symlinks? If so that may explain problems. Symlinks do not properly work, see https://github.com/rsyslog/rsyslog/issues/3550



> Will that configuration work stable with wildcards and polling mode? What should we use to have production stable monitoring of files, _inotify + statefilealwaysremove_?

I think it should (but I had only a very brief look plus it is only a snippet). Polling has some higher latency than inotify, so it may miss updates (file being recreated twice within the polling interval). Other than that, the modes are equivalent.

Do you rotate the files? If so, how?

**Comment by waster** [comment]
No, we don't use symlinks at all. Log rotation is done internally with Icecast itself when _logsize_ setting in configuration file is configured. 

So will _inotify + statefilealwaysremove_ imfile options help us to provide consistent and stable imfile work in that case?

**Comment by rgerhards** [comment]
statefilealwaysremove does not exist. It was part of a PR that did not work.

So this looks like an issue. We probably need a debug log to see what happens.

**Comment by rgerhards** [comment]
It would help to know how overcast actually rotates the log.

**Comment by waster** [comment]
Problem is that it is rarely happen, more often in inotify mode. I'll try to enable debug with rsyslog and increase a frequency of the icecast log rotations.

**Comment by rgerhards** [comment]
Sorry for the very late reply. This is probably no longer relevant to the original setup, but I want to leave the current triage result for people who find this via search. We do not lightly close old issues or close them just because of age; we are working through the backlog and checking what can still be acted on.

I am closing this issue because we cannot reproduce or debug it further from the information available here. The report was for rsyslog 8.1903 on Ubuntu 16.04, and the last actionable state was that a debug log and more detail about Icecast's rotation behavior were needed. There has not been enough information to tie the behavior to a current imfile defect.

For current deployments, use current rsyslog and current imfile syntax. Polling mode can work with wildcards, but it can miss very fast create/rotate/recreate cycles inside the polling interval. Inotify usually gives lower latency, but rotation details still matter. Symlinks and unusual rename/truncate behavior are especially important to call out.

If this problem still occurs on a supported current version, please create a new issue with:

- the exact rsyslog version and platform
- the full imfile input configuration
- whether the application renames, truncates, deletes, or recreates the log file
- a minimal rotation reproducer if possible
- rsyslog debug output around the point where the file stops being followed

That will make it possible to investigate against the current imfile code instead of guessing from an old unreproducible state.
